src/Surfnet/StepupGateway/GatewayBundle/Saml/Proxy/ProxyStateHandler.php

```diff
@@ -330,13 +331,30 @@ public function getPreferredLocale()
         return $this->get('locale');
     }
 
+    /**
+     * note that the authentication mode is stored outside the session path, to enable other state handlers
+     * to retrieve the Authentication state for a given authentication request id.
+     *
+     * @param $requestId
+     * @param $authenticationMode
+     */
+    public function markAuthenticationModeForRequest($requestId, $authenticationMode)
+    {
+        $this->session->set('surfnet/gateway/auth_mode/' . $requestId, $authenticationMode);
```

Comment:
This will make all code very hard to maintain. Personally I don't know how state is managed and when and how that effects the state of the session in terms of the two esponseContexts, the two ProxyStateHandlers their relation and how and when they interact with session data.

Certainly when dynamic session paths are used and a session path is injected with DI but internally another session path is hardcoded.

Reply:
Yeah, this is a solution to the GSSP problem, where you would forget your authentication mode in between the GSSP callout. The session path is not dynamic perse. The path is simply a prefix to the session key. I've made that explicitly different from the on the state handler would use as we need to read this session key from all different handlers.

This problem would instantly be resolved when state handling was done on the request-id in the first place.

Definitely one to fix-up when refactoring the state handling in gateway
https://www.pivotaltracker.com/story/show/172196431